packages/mcp/src/index.ts

```diff
@@ -543,22 +824,39 @@ export class MCPProxy {
             version: '1.0.0',
           });
 
-          const transport = new PrefixedStdioClientTransport(
+          // Resolve environment variables using the new secret system
+          const resolvedEnv = await this.resolveServerEnvironment(targetServer);
+
+          const transport = new ReconnectablePrefixedStdioClientTransport(
             targetServer.name,
             {
               command: targetServer.command,
               args: targetServer.args || [],
-              env: { ...process.env, ...targetServer.env } as Record<
-                string,
-                string
-              >,
+              env: resolvedEnv,
+              reconnection: {
+                initialDelay: 1000,
+                maxDelay: 30000,
+                backoffMultiplier: 2,
+                maxRetries: 10,
+                jitter: 0.25,
+              },
+              healthChecks: true,
+              healthCheckInterval: 30000,
             },
           );
 
+          // Store transport for management
+          this._transports.set(targetServer.name, transport);
+
```

Comment:
When a transport reconnects we arrive here with `stateChange.to === ConnectionState.Connected`, but `targetServer` was retrieved from `this.connectedServers` *before* we removed the server in the disconnect branch. After the first disconnect that map no longer contains the entry, so on later reconnect notifications `targetServer` is `undefined` and `handleServerConnection` never runs. The proxy keeps the server in `disconnectedServers`, tools never get rediscovered, and the proxy continues treating the server as offline even though the process is back. Please grab the definition from `disconnectedServers` (and clear the transient error) when the connected map misses so reconnection can actually restore state.